R/gh_token.R

```diff
@@ -164,21 +180,162 @@ gh_auth <- function(token) {
+}
+
+# inlined from credentials and generalized to non-github.com URL
+# is_interactive() is an unexported helper defined in utils.R
+# morally, is rlang::is_interactive()
+set_github_pat2 <- function(api_url = default_api_url(),
+                            force_new = FALSE,
+                            validate = is_interactive(),
+                            verbose = validate) {
+  if (!can_load("credentials")) {
+    return(FALSE)
+  }
+
+  host_url <- get_hosturl(api_url)
+  prot_host <- strsplit(host_url, "://", fixed = TRUE)[[1]]
+  if (is_github_dot_com(api_url)) {
+    # what's currently in credentials
+    # pat_user <- Sys.getenv("GITHUB_PAT_USER", 'PersonalAccessToken')
+    # TODO: I propose we change the user env var to GITHUB_USER
+    user_env_var <- "GITHUB_PAT_USER"
+    token_env_var <- "GITHUB_PAT"
+  } else {
+    user_env_var <- paste0("GITHUB_USER_", slugify_url(api_url))
+    token_env_var <- paste0("GITHUB_PAT_", slugify_url(api_url))
+  }
+  api_url <- get_apiurl(api_url)
+
+  # The username doesn't have to be real, Github seems to ignore username for PATs
+  pat_user <- Sys.getenv(user_env_var, "PersonalAccessToken")
+  # credentials does: pat_url <- sprintf("https://%s@github.com", pat_user)
+
+  pat_url <- sprintf(
+    "%s://%s@%s",
+    prot_host[[1]], pat_user, prot_host[[2]]
+  )
+
+  if (isTRUE(force_new)) {
+    credentials::git_credential_forget(pat_url)
+  }
+
+  # credentials does the "hack" unconditionally, but I think this is better
+  if (is_interactive()) {
+    if(isTRUE(verbose)) {
+      message2("If prompted for GitHub credentials, enter your PAT in the password field")
+    }
+    askpass <- Sys.getenv('GIT_ASKPASS')
+    if (nchar(askpass)) {
+      # Hack to override prompt sentence to say "Token" instead of "Password"
+      withr::local_envvar(c(
+        GIT_ASKTOKEN = askpass,
+        GIT_ASKPASS = system.file(
+          "ask_token.sh", package = "gh", mustWork = TRUE
+        ),
+        GIT_ASKTOKEN_NAME = sprintf(
+          "Personal Access Token (PAT) for %s", host_url
+        )
+      ))
+    }
+  } else {
+    withr::local_envvar(c(
+      GIT_ASKPASS = system.file(
+        "empty_token.sh", package = "gh", mustWork = TRUE
+      )
+    ))
+  }
+
+  for(i in 1:3) {
+    cred <- credentials::git_credential_ask(pat_url, verbose = verbose)
+    if (length(cred$password) && !is.na(cred$password)) {
+      if (nchar(cred$password) < 40) {
+        message2("Please enter a token in the password field, not your master password! Let's try again :-)")
+        message2(sprintf(
+          "To generate a new token, visit: %s/settings/tokens",
+          host_url
+        ))
+        credentials::credential_reject(cred)
+        next
+      }
+      if (isTRUE(validate)) {
+        hx <- curl::handle_setheaders(curl::new_handle(), Authorization = paste("token", cred$password))
+        req <- curl::curl_fetch_memory(sprintf("%s/user", api_url), handle = hx)
```

Comment:
This can be done with gh technology if this stays here vs. using curl directly.

Reply:
I am not sure if we need to validate the token. The real query might fail, anyway, the token can be invalidated externally, etc. so errors must be handled there.

Reply:
I am just copying what `credentials::set_github_pat()` does. I can provide some context, though.

This is a faithful implementation of the official workflow recommended in "Typical Use of Git Credential":

https://git-scm.com/docs/git-credential#_typical_use_of_git_credential

It makes sense to validate right after the `git credential fill` call, which happens inside `credentials::git_credential_ask(pat_url, verbose = verbose)` for us.  That fetches a matching credential from the store or prompts the user for the PAT. Then you immediately do either `git credential approve` (valid PAT)  or `git credential reject` (invalid PAT). Implications:

* New PAT we just from user: By validating, we make sure we don't put a bad PAT into the store.
* Existing PAT we got from the user If PAT is good, it persists, but if it's known to have gone bad, we remove it.

This seems worthwhile to me.

Yeah, it's also true that a subsequent request can fail, but remember these are all happening extremely close in time. This validation only happens once per session, when we move the PAT from the store to an env var. Token invalidation between pulling from the store and usage seems unlikely. More likely is the token having the wrong scopes vis-a-vis a particular request.